code/modules/maps/template_types/random_exoplanet/random_planet_level_data.dm

```diff
@@ -0,0 +1,143 @@
+
+///Level data for generating surface levels on exoplanets
+/datum/level_data/planetoid/exoplanet
+	base_area = /area/exoplanet
+	base_turf = /turf/exterior/dirt
+
+///Level data for generating underground levels on exoplanets
+/datum/level_data/planetoid/exoplanet/underground
+	base_area = /area/exoplanet/underground
+	base_turf = /turf/exterior/volcanic
+	level_generators = list(
+		/datum/random_map/noise/exoplanet/mantle,
+	)
+
+/datum/level_data/planetoid/initialize_level_id()
+	if(level_id)
+		return
+	level_id = "planetoid_[level_z]_[sequential_id(/datum/level_data)]"
+
+///Prepare our level for generation/load. And sync with the planet template
+/datum/level_data/planetoid/before_template_generation(datum/map_template/template, var/datum/planetoid_data/gen_data)
+	//In cases when the planetoid_data isn't shared with us, like from loading non-random planet map templates, we can try fetching it.
+	if(!gen_data)
+		gen_data = SSmapping.planetoid_data_by_z[level_z]
+	if(!gen_data)
+		CRASH("'[src]', id:'[level_id]', z:'[level_z]'. Couldn't access expected planetoid_data!")
+
+	//Apply parent's prefered bounds if we don't have any preferences
+	if(!level_max_width && gen_data.width)
+		level_max_width = gen_data.width
+	if(!level_max_height && gen_data.height)
+		level_max_height = gen_data.height
+
+	//Refresh bounds
+	setup_level_bounds()
+
+	//override atmosphere
+	apply_planet_atmosphere(gen_data)
+
+	//Rename main area and level
+	adapt_location_name(gen_data.name)
+
+	//Try to adopt our parent planet's ambient lighting preferences
+	apply_planet_ambient_lighting(gen_data)
+
+///If we're getting atmos from our parent planet, decide if we're going to apply it, or ignore it
+/datum/level_data/planetoid/proc/apply_planet_atmosphere(var/datum/planetoid_data/P)
+	if(istype(exterior_atmosphere))
+		return //level atmos takes priority over planet atmos
+	exterior_atmosphere = P.atmosphere.Clone() //Make sure we get one instance per level
+
+///Apply our parent planet's ambient lighting settings if we want to.
+/datum/level_data/planetoid/proc/apply_planet_ambient_lighting(var/datum/planetoid_data/P)
+	if(!ambient_light_level)
+		ambient_light_level = P.surface_light_level
+	if(!ambient_light_color)
+		ambient_light_level = P.surface_light_color
+
+/datum/level_data/planetoid/adapt_location_name(location_name)
+	if(!(. = ..()))
+		return
+	if(!ispath(base_area) || ispath(base_area, world.area))
+		return
+	var/area/A = get_base_area_instance()
+	//Make sure we're not going to rename the world's base area
+	if(!istype(A, world.area))
+		global.using_map.area_purity_test_exempt_areas |= A.type //Make sure we add any of those, so unit tests calm down when we rename
+		A.SetName("[location_name]")
+
+///Try to spawn the given amount of ruins onto our level. Returns the template types that were spawned
+/datum/level_data/planetoid/proc/seed_ruins(var/budget = 0, var/list/potentialRuins)
+	//#TODO: Fill in allowed area from a proc or something
```

Comment:
```suggestion
	if(!length(potentialRuins))
		log_world("Ruin loader was given no ruins to pick from.")
		return list()
	//#TODO: Fill in allowed area from a proc or something
```
This guards against `pick(candidates_ruins)` runtiming with `pick from empty list` if `potentialRuins` is empty.